permissions/views.py

```diff
@@ -0,0 +1,204 @@
+
+    return render(request, "team_page.html", context)
+
+
+@require_POST
+@login_required
+def create_role(request):
+    team_id = request.POST.get("team_id")
+    role_name = request.POST.get("role_name")
+    members_list = request.POST.getlist("members")
+
+    try:
+        team = Team.objects.get(id=team_id)
+        role = Role(
+            name=role_name,
+            created_by=request.user,
+            team=team
+        )
+
+        role.save()
+
+        members = User.objects.filter(username__in=members_list)
+        role.members.add(*members)
+
+        role.save()
+    except Team.DoesNotExist:
+        return redirect('permissions:home')
+
+    return redirect('permissions:team_detail', team_id)
+
+
+@require_POST
+@login_required
+def add_permission(request):
+    team_id = request.POST.get("team_id")
+    course_id = request.POST.get("courses")
+    role_id = request.POST.get("role")
+    perm_type = request.POST.get("perm_type")
+    units = request.POST.getlist("units")
+
+    team = Team.objects.get(id=team_id)
+    course = Course.objects.get(id=course_id)
+    role = Role.objects.get(id=role_id)
+
+    # Create permission obj for each unit
+    for unit in units:
+        type, id = unit.split("_")
+
+        if type == "quiz":
+            content_object = Quiz.objects.get(pk=id)
+        else:
+            content_object = Lesson.objects.get(pk=id)
+
+        permission = Permission(
+            team=team,
+            course=course,
+            perm_type=perm_type,
+            content_object=content_object
+        )
+
+        permission.save()
+
+        permission.role.add(role)
+
+        permission.save()
+
+    return redirect('permissions:team_detail', team_id)
+
+
+@login_required
+def delete_permission(request, permission_id, team_id):
+    '''
+    Delete permission.
+    Only team creator can delete
+    '''
+
+    try:
+        team = Team.objects.get(pk=team_id)
+
+        if team.created_by == request.user:
+            Permission.objects.get(pk=permission_id).delete()
+
+            return redirect('permissions:team_detail', team_id)
+    except Team.DoesNotExist:
+        return redirect('permissions:home')
+
+
+@login_required
+def get_modules(request):
+    ''' Get modules belonging to a course '''
+
+    course_id = request.GET.get("course_id")
+    course = get_object_or_404(Course, pk=course_id)
+    modules = course.learning_module.all()
+
+    units = []
+
+    for module in modules:
+        quiz_units = module.get_quiz_units()
+        lesson_units = module.get_lesson_units()
+
+        quiz_units_data = [
+            {"key": "quiz_{}".format(quiz_unit.id),
+             "name": quiz_unit.description} for quiz_unit in quiz_units
+        ]
+
+        lesson_units_data = [
+            {"key": "lesson_{}".format(lesson_unit.id),
+             "name": lesson_unit.name} for lesson_unit in lesson_units
+        ]
+
+        units = quiz_units_data + lesson_units_data
```

Comment:
Do you need this extra list quiz_units_data and lesson_units_data? you can directly send all the units.

Reply:
Yeah, I need the formatted key, so that i can refer what type of unit it is in the frontend. 

> Also instead of + you can use append

I'll change this.

Reply:
Yeah, I need the formatted key, so that i can refer what type of unit it is in the frontend
- In that case create a model method which returns specific data that you need. Don't get unnecessary data from database if you are not going to use it.

+ works fine. Is there any reason not to use it ?
- Sorry I got it wrong instead of `append` it is `extend`. `+` is slow as compared to append/extend.

Reply:
> create a model method

I do not think that the format of data used in the templates should drive the model methods, model methods should be driven by the API that you think that the application should have.

Anyway, in that case you can avoid ```get_lesson_units```and ```get_quiz_units``` and directly query modules (using the ```get_all_units``` method) and fetch desired values only using values_list

Once changed you will not need to use extend or concatenation of any kind.